renderdoc/driver/vulkan/vk_manager.cpp

```diff
@@ -245,6 +245,155 @@ void VulkanResourceManager::MergeBarriers(rdcarray<rdcpair<ResourceId, ImageRegi
   TRDBG("Post-merge, there are %u states", (uint32_t)dststates.size());
 }
 
+#if ENABLED(RDOC_NEW_IMAGE_STATE)
+template <typename SerialiserType>
+void VulkanResourceManager::SerialiseImageStates2(SerialiserType &ser,
+                                                  std::map<ResourceId, LockingImageState> &states)
+{
+  SERIALISE_ELEMENT_LOCAL(NumImages, (uint32_t)states.size());
+
+  auto srcit = states.begin();
+
+  std::set<ResourceId> updatedState;
+
+  for(uint32_t i = 0; i < NumImages; i++)
+  {
+    SERIALISE_ELEMENT_LOCAL(Image, (ResourceId)(srcit->first)).TypedAs("VkImage"_lit);
+    if(ser.IsWriting())
+    {
+      LockedImageStateRef lockedState = srcit->second.LockWrite();
+      ::ImageState &ImageState = *lockedState;
+      SERIALISE_ELEMENT(ImageState);
+      ++srcit;
+    }
+    else
+    {
+      ImageState imageState;
+
+      if(ser.VersionLess(0x11))
+      {
+        ImageLayouts imageLayouts;
+        {
+          ImageLayouts &ImageState = imageLayouts;
+          SERIALISE_ELEMENT(ImageState);
+        }
+        if(IsReplayingAndReading())
+        {
+          imageState = ImageState(VK_NULL_HANDLE, imageLayouts.imageInfo);
+
+          rdcarray<TaggedImageSubresourceState> subresourceStates;
+          subresourceStates.reserve(imageLayouts.subresourceStates.size());
+
+          for(ImageRegionState &st : imageLayouts.subresourceStates)
+          {
+            TaggedImageSubresourceState p;
+            p.range = st.subresourceRange;
+            p.state.oldQueueFamilyIndex = st.dstQueueFamilyIndex;
+            p.state.newQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+            p.state.oldLayout = st.newLayout;
+            p.state.newLayout = imageState.GetImageInfo().initialLayout;
+            p.state.refType = eFrameRef_Maximum;
+            subresourceStates.push_back(p);
+          }
+
+          imageState.subresourceStates.FromArray(subresourceStates);
+        }
+      }
+      else
+      {
+        {
+          ::ImageState &ImageState = imageState;
+          SERIALISE_ELEMENT(ImageState);
+        }
+        if(IsReplayingAndReading())
+        {
+          imageState.newQueueFamilyTransfers.clear();
+          for(auto it = imageState.subresourceStates.begin();
+              it != imageState.subresourceStates.end(); ++it)
+          {
+            // Set the current image state (`newLayout`, `newQueueFamilyIndex`, `refType`) to the
+            // initial image state, so that calling `ResetToOldState` will move the image from the
+            // initial state to the state it was in at the beginning of the capture.
+            ImageSubresourceState &state = it->state();
+            state.newLayout = imageState.GetImageInfo().initialLayout;
+            state.newQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+          }
+        }
+      }
+      if(HasLiveResource(Image))
+      {
+        ResourceId liveid = GetLiveID(Image);
+        if(liveid != ResourceId())
```

Comment:
This was based on the same pattern in `SerialiseImageStates`, which checks both `HasLiveResource(Image)` and `liveid != ResourceId()`. It looks like `HasLiveResource` only checks that `Image` is in `m_LiveResourceMap`, but doesn't verify that the liveid is non-nil.

It seems plausible that `m_LiveResourceMap` should never have nil live IDs, so I'm happy to remove this secondary check if you think it's unnecessary.

Reply:
Yeh it's a little messy in that there are multiple parallel maps expected to be updated in sync - so if there's a live resource in `m_LiveResourceMap` then it should also have a live ID. At the least if that is violated then the bug is not in this code.

I think `SerialiseImageStates` only does both checks because the code has changed and previously those two checks were entirely separate, it should be safe to drop the second if.